template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmTemplateProcessor.java

```diff
@@ -186,6 +187,20 @@ public void addVariables(List<ApiClusterTemplateVariable> vars) {
         }
     }
 
+    public void extendTemplateWithAdditionalServices(Map<String, ApiClusterTemplateService> hostGroupServices) {
```

Comment:
The current signature is fine for Knox, but for more complex services, we may need to create "fake" services to assign different roles to different host groups.  (Eg. if you wanted to add a server to master group and client to all other groups.)

I think it would be cleaner for the config provider to generate any number of services, and map each role config to one or more host groups explicitly.

Reply:
Besides Knox I don't think we need to add new services to the blueprint dynamically. If the requirements change we can adjust the code. I don't want it for too general now as it is complicated to handle all the scenarios.